src/main/scala/calc/Compiler.scala

```diff
@@ -61,12 +64,34 @@ object Compiler {
     ir.Hashers.hashClassDef(classDef)
   }
 
+
+
+  def findCapture(tree: Tree): Set[String] = {
+    tree match {
+      case _:Literal => Set()
+      case BinaryOp(_, l, r) => findCapture(l) ++ findCapture(r)
+      case Let(i, v, body) => findCapture(body) ++ findCapture(v) - i.name
+      case Ident(n) => Set(n)
+      case If(cond, thenp, elsep) => findCapture(cond) ++ findCapture(thenp) ++ findCapture(elsep)
+      case Closure(idents, body) => findCapture(body) -- idents.toSet.flatMap((x:Tree) => findCapture(x))
+      case Call(f, args) => findCapture(f) ++ args.toSet.flatMap((x:Tree) => findCapture(x))
```

Comment:
Calling `args.toSet` is weird, because having a `Set[Tree]` is weird. `Tree`s are not meant to be compared for equality, and especially their hash code is expensive to compile (it has to walk the entire subtree). You should first `flatMap` *then* `.toSet`, so that you never have a transient `Set[Tree]` (only `Set[String]`).